### PR TITLE
Improved units processing performance

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
Related to IDAES/idaes-pse#132: Pyomo's units processing entailed a significant performance overhead.  This PR switches the internal units processing to operate *only* using raw `pint` units (and not pint/pyomo object tuples).  This also moves to actively converting pyomo unit expressions to simple compound pint unit objects.  Finally, it updates some of the string representations to output pint strings that are closer to teh compact style used by Pyomo expressions.

The net effect (for IDAES models) is to speed up model creation time by almost 50%.

## Changes proposed in this PR:
- Convert unit extraction visitor to only process pint units
- Convert component unit expressions to a single Pyomo unit object (erapping a compound pint unit)
- Update the string representation of compound pint units to match pyomo expressions

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by blnicho — APPROVED** [review]
A few questions about commented code but otherwise I think this looks fine.

File: pyomo/util/check_units.py [hunk 1]
```diff
@@ -65,14 +66,22 @@ def assert_units_equivalent(*args):
     ------
     :py:class:`pyomo.core.base.units_container.UnitsError`, :py:class:`pyomo.core.base.units_container.InconsistentUnitsError`
     """
+    # pyomo_unit_compare, pint_unit_compare = units._get_units_tuple(args[0])
+    # for expr in args[1:]:
+    #     # this call will raise an exception if an inconsistency is found
+    #     pyomo_unit, pint_unit = units._get_units_tuple(expr)
+    #     if not UnitExtractionVisitor(units)._pint_units_equivalent(pint_unit_compare, pint_unit):
+    #         raise UnitsError \
+    #             ("Units between {} and {} are not consistent.".format(str(pyomo_unit_compare), str(pyomo_unit)))
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/util/check_units.py]: Yes. Done.

File: pyomo/util/check_units.py [hunk 2]
```diff
@@ -25,6 +25,7 @@
 from pyomo.core.expr.template_expr import IndexTemplate
 from pyomo.core.expr.numvalue import native_types
 from pyomo.util.components import iter_component
+#from pyomo.core.base.units_container import UnitExtractionVisitor
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/util/check_units.py]: Yes. Done.

**Review by jsiirola — COMMENTED** [review]
Resolved @blnicho comments.

**Review by carldlaird — APPROVED** [review]
This looks reasonable to me. At some point, we may want to rethink "get_units" on indexed components - in case we later support different units on different indices.

**Comment by jsiirola** [comment]
@carldlaird I agree.  I will look shortly and see if we can remove `get_units` from the containers... if we can (and tests pass) I will add it to this PR.

**Comment by jsiirola** [comment]
@carldlaird: the problem I encounteres was working with templated expressions: consider the following test:
```python
model.S = Set()
i = EXPR.IndexTemplate(model.S)
j = EXPR.IndexTemplate(model.S)
self._get_check_units_ok(i, uc, None, EXPR.IndexTemplate)

model.mat = Var(model.S, model.S)
self._get_check_units_ok(model.mat[i,j+1], uc, None, EXPR.GetItemExpression)
```
In the last line, we are walking an expression tree where one of the child nodes is the *indexed* var.  Right now, this works because the units are consistent across all the _VarData, so it is safe to return an answer.

For the time being, I think we will leave things as they are and revisit it again if we have a need to have heterogeneous units within containers.

**Pull request merged**